Docs: Containerized plugins usage

tomhjp opened the pull request:
Adds usage documentation for the new containerized plugins feature coming in 1.15.0. API documentation to follow separately.

swenson commented on website/content/docs/commands/plugin/register.mdx:
```diff
@@ -44,14 +44,25 @@ flags](/vault/docs/commands) included on all commands.
 ### Command options
 
-- `-sha256` `(string: <required>)` - Checksum (SHA256) of the plugin binary.
+- `-sha256` `(string: <required>)` - SHA256 of the plugin binary or the oci_image
```
```suggestion
- `-sha256` `(string: <required>)` - SHA256 of the plugin binary or the OCI image
```

swenson commented on website/content/docs/commands/plugin/register.mdx:
```diff
@@ -44,14 +44,25 @@ flags](/vault/docs/commands) included on all commands.
-- `-command` `(string: "")` - Name of the command to run to invoke the binary.
-  By default, this is the name of the plugin.
+- `-command` `(string: "")` - Command to spawn the plugin. This defaults to the
+  name of the plugin if both oci_image and command are unspecified.
```
```suggestion
  name of the plugin if both `-oci_image` and `-command` are unspecified.
```

swenson commented on website/content/docs/commands/plugin/register.mdx:
```diff
@@ -44,14 +44,25 @@ flags](/vault/docs/commands) included on all commands.
+  args, and env will update the container's entrypoint, args, and environment
+  variables (append-only) respectively.
+
+- `-runtime` `(string: "")` - Vault plugin runtime to use if -oci_image is
```
```suggestion
- `-runtime` `(string: "")` - Vault plugin runtime to use if `-oci_image` is
```

swenson commented on website/content/docs/commands/plugin/register.mdx:
```diff
@@ -44,14 +44,25 @@ flags](/vault/docs/commands) included on all commands.
+- `-runtime` `(string: "")` - Vault plugin runtime to use if -oci_image is
+  specified.
+
+- `-version` `(string: "")` - Semantic version of the plugin. Used as the tag
+  when specifying oci_image, but with any leading 'v' trimmed.
```
Do you mean that any leading `v` will be trimmed automatically, or that you should not specify the `v`?

```suggestion
- `-version` `(string: "")` - Semantic version of the plugin. Used as the tag
  when specifying `-oci_image`, but with any leading 'v' trimmed.
```

swenson commented on website/content/docs/commands/plugin/runtime/deregister.mdx:
```diff
@@ -0,0 +1,31 @@
+Deregister an existing plugin runtime in the catalog with the given name. If
+any registered plugin references the plugin runtime, an error is returned. If
+the plugin runtime does not exist, an error is returned. The `-type` flag
+currently only accepts "container".
```
```suggestion
currently only accepts `container`.
```

swenson commented on website/content/docs/commands/plugin/runtime/deregister.mdx:
```diff
@@ -0,0 +1,31 @@
+flags](/vault/docs/commands) included on all commands.
+
+- `-type` `(string: "")` - Plugin runtime type. Vault currently only supports
+  "container" runtime type.
```
```suggestion
  `container` runtime type.
```

swenson commented on website/content/docs/commands/plugin/runtime/deregister.mdx:
```diff
@@ -0,0 +1,31 @@
+The following flags are available in addition to the [standard set of
+flags](/vault/docs/commands) included on all commands.
+
+- `-type` `(string: "")` - Plugin runtime type. Vault currently only supports
```
Is this required?

swenson commented on website/content/docs/commands/plugin/runtime/index.mdx:
```diff
@@ -0,0 +1,69 @@
+
+# plugin runtime
+
+The "plugin runtime" command groups subcommands for interacting with
```
```suggestion
The `plugin runtime` command groups subcommands for interacting with
```

swenson commented on website/content/docs/commands/plugin/runtime/index.mdx:
```diff
@@ -0,0 +1,69 @@
+To use a registered plugin runtime, use the `-runtime` option with the
+[plugin registration command](/vault/docs/commands/plugin/register).
+
+Currently, the only supported type of plugin runtime is "container".
```
```suggestion
Currently, the only supported type of plugin runtime is `container`.
```

swenson commented on website/content/docs/commands/plugin/runtime/register.mdx:
```diff
@@ -0,0 +1,55 @@
+# plugin runtime register
+
+The `plugin runtime register` command registers a new plugin runtime in Vault's plugin runtime catalog.
+Vault currently only supports plugin runtimes of type "container".
```
```suggestion
Vault currently only supports plugin runtimes of type `container`.
```

swenson commented on website/content/docs/commands/plugin/runtime/register.mdx:
```diff
@@ -0,0 +1,55 @@
+  Defaults to no limit.
+
+- `-oci_runtime` `(string: "")` - OCI-compliant container runtime to use.
+  Default is "runsc", gVisor's OCI runtime.
```
```suggestion
  Default is `runsc`, gVisor's OCI runtime.
```

swenson commented on website/content/docs/commands/plugin/runtime/register.mdx:
```diff
@@ -0,0 +1,55 @@
+### Command options
+
+- `-type` `(string: <required>)` - Plugin runtime type. Vault currently only
+  supports "container" as a runtime type.
```
```suggestion
  supports `container` as a runtime type.
```

swenson commented on website/content/docs/commands/plugin/runtime/register.mdx:
```diff
@@ -0,0 +1,55 @@
+- `-cgroup_parent` `(string: "")` - Parent cgroup to set for each container.
+  This can be used to control the total resource usage for a group of plugins.
+
+- `-cpu_nanos` `(int: 0)` - CPU limit to set per container in nanos. Defaults
```
It might be good to clarify here what a "nano" is? I presumed a billionth of a CPU core.

Possibly out the scope of this PR, but maybe it might be more useful to support Kubernetes-like unit suffixes (for here and for memory), so that people could use the more familiar `250m` for a quarter CPU instead of `250000`.

tomhjp replied:
Thanks, clarified the language to billionths of a CPU core. And agreed on the unit suffixes - I'm going to try to do that for 1.15.1

swenson commented on website/content/docs/commands/plugin/runtime/list.mdx:
```diff
@@ -0,0 +1,38 @@
+### Output options
+
+- `-format` `(string: "table")` - Print the output in the given format. Valid
+  formats are "table", "json", or "yaml". This can also be specified via the
```
```suggestion
  formats are `table`, `json`, or `yaml`. This can also be specified via the
```

swenson commented on website/content/docs/commands/plugin/runtime/list.mdx:
```diff
@@ -0,0 +1,38 @@
+### Command options
+
+- `-type` `(string: "")` - Plugin runtime type. Vault currently only supports
+  "container" runtime type.
```
```suggestion
  `container` runtime type.
```

swenson commented on website/content/docs/commands/plugin/runtime/index.mdx:
```diff
@@ -0,0 +1,69 @@
+Vault's plugin runtimes, configuring the context within which external
+plugins can be run.
+
+To use a registered plugin runtime, use the `-runtime` option with the
```
It might be worth commenting about common ones being `runc` and `runsc`, what they are, and if they are available by default without registering, or if they need to be registered first?

tomhjp replied:
Yeah good idea. Added a paragraph on this below.

swenson commented on website/content/docs/commands/plugin/runtime/index.mdx:
````diff
@@ -0,0 +1,69 @@
+```text
+Usage: vault plugin runtime <subcommand> [options] [args]
+
+  # ...
````
If this `# ...` correct, or are you just abbreviating?

swenson commented on website/content/docs/commands/plugin/runtime/info.mdx:
```diff
@@ -0,0 +1,48 @@
+
+Displays information about a plugin runtime in the catalog with the given name. If
+the plugin runtime does not exist, an error is returned. The `-type` flag currently
+only accepts "container".
```
```suggestion
only accepts `container`.
```

swenson commented on website/content/docs/commands/plugin/runtime/info.mdx:
```diff
@@ -0,0 +1,48 @@
+  processes.
+
+- `-format` `(string: "table")` - Print the output in the given format. Valid
+  formats are "table", "json", or "yaml". This can also be specified via the
```
```suggestion
  formats are `table`, `json`, or `yaml`. This can also be specified via the
```

swenson commented on website/content/docs/commands/plugin/runtime/info.mdx:
```diff
@@ -0,0 +1,48 @@
+### Command options
+
+- `-type` `(string: "")` - Plugin runtime type. Vault currently only supports
+  "container" runtime type.
```
```suggestion
  `container` runtime type.
```

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+description: Running external plugins for Vault in containers.
+---
+
+# Containerized Plugins (Beta)
```
I'm not sure if we usually include the beta in the page title, but more as a note in the body?

fairclothjm replied:
```suggestion
# Containerized plugins (Beta)
```

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+  Use the `DOCKER_HOST` environment variable to configure a custom address.
+
+* (Optional) gVisor installed and configured as a runtime. Vault uses gVisor's
+  runsc container runtime by default. However, it can also be configured to use
+  other runtimes such as Docker's default, runc. In addition to the
+  [official installation instructions](https://gvisor.dev/docs/user_guide/install/),
+  runsc requires an additional runtime argument to enable Unix domain socket
```
```suggestion
  `runsc` container runtime by default. However, it can also be configured to use
  other runtimes such as Docker's default, `runc`. In addition to the
  [official installation instructions](https://gvisor.dev/docs/user_guide/install/),
  `runsc` requires an additional runtime argument to enable Unix domain socket
```

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+    * [Docker](https://docs.docker.com/engine/install/)
+    * Or [rootless Docker](https://docs.docker.com/engine/security/rootless/)
+
+2. [Install runsc](https://gvisor.dev/docs/user_guide/install/)
```
```suggestion
2. [Install `runsc`](https://gvisor.dev/docs/user_guide/install/)
```

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+
+2. [Install runsc](https://gvisor.dev/docs/user_guide/install/)
+
+3. Configure runsc to allow Unix sockets between the host and plugin, updating
+   the runsc install location if you didn't use the default:
```
```suggestion
3. Configure `runsc` to allow Unix sockets between the host and plugin, updating
   the `runsc` install location if you didn't use the default:
```

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    ```bash
+    git clone https://github.com/hashicorp/vault-plugin-secrets-kv.git
+    cd vault-plugin-secrets-kv/
+    GOOS=linux go build -o kv cmd/vault-plugin-secrets-kv/main.go
````
`CGO_ENABLED=0` ?

swenson commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+### Container lifecycle and metadata
+
+Like any other external plugin, Vault will automatically manage the lifecycle
+of plugin containers. If they are killed out of band, Vault will restart it
+before servicing any requests that need to be handled by it. Vault will also
```
```suggestion
of plugin containers. If they are killed out of band, Vault will restart them
before servicing any requests that need to be handled by them. Vault will also
```

swenson reviewed (approved):
LGTM. Mostly some style nits, and a small couple of questions.

fairclothjm commented on website/content/docs/commands/plugin/register.mdx:
```diff
@@ -44,14 +44,25 @@ flags](/vault/docs/commands) included on all commands.
+  starting. This flag can be specified multiple times to specify multiple
+  environment variables.
+
+- `-oci_image` `(string: "")` - OCI image to run. If specified, setting command,
+  args, and env will update the container's entrypoint, args, and environment
+  variables (append-only) respectively.
```
Based on @swenson 's suggestions:

```suggestion
- `-oci_image` `(string: "")` - OCI image to run. If specified, setting `-command`,
  `-args`, and `-env` will update the container's entrypoint, args, and environment
  variables (append-only) respectively.
```

fairclothjm commented on website/content/docs/commands/plugin/runtime/list.mdx:
```diff
@@ -0,0 +1,38 @@
+# plugin list
+
+Lists available plugin runtimes registered in the catalog. This does not list whether
+plugin runtimes are in use, but rather just their availability.
```
```suggestion
plugin runtimes are in use, but rather just their availability. If
there are no registered plugin runtimes, an error is returned.
```

fairclothjm replied:
We had the same note for `runtime info` so I thought we may want to add here as well.

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    ENTRYPOINT [ "/bin/kv" ]
+    ```
+
+1. Build the image:
+    ```bash
+    docker build -t hashicorp/vault-plugin-secrets-kv .
+    ```
````
````suggestion
1. Build the image:
    ```shell-session
    docker build -t PATH/TO/THE/PLUGIN/BINARY .
    ```
````

I'm not entirely clear from the previous steps what the build path is supposed to be. It it a path to the compiled binary or something else? For what it's worth, this is why generalizing is useful for example code, it helps folks note how the commands are related :)

Style correction: use `shell-session` for bash code, generalize example commands when possible

fairclothjm replied:
@schavis `hashicorp/vault-plugin-secrets-kv` is the "tag" being given to the image by the `docker build -t` command (-t for tag). The tag is a human-readable ID given to the image. The build path is specified by the `.` which means the current working directory.

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    docker build -t hashicorp/vault-plugin-secrets-kv .
+    ```
+
+The image entrypoint, command, and environment can be further configured at
+registration time by using the `-command`, `-args`, and `-env` flags for
+`vault plugin register`.
+
+### Registration
````
```suggestion
## Step 3: Register the plugin

<Tip>

Use the `-command`, `-args`, and `-env` flags to further customize the entrypoint,
command, and environment for your plugin.

</Tip>

```

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+
+### Registration
+
+Once Vault has a suitable container runtime available, the remaining steps are
+very similar to registering and using other external plugins. The config setting
+for [`plugin_directory`](/vault/docs/configuration#plugin_directory) is not required
+for containerized plugins, but any plugin containers must already be available
+locally when mounted.
```
```suggestion
Once you containerize the plugins, registering them with Vault is similar to
registering any other external plugins as long as the containers are available
locally when mounted.
```
Focus on what is required rather than what is not required since we're outlining steps the reader should take

Style correction: write in active voice,

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+locally when mounted.
+
+1. Store the SHA256 of the plugin image:
+    ```bash
+    SHA256="$(docker images --no-trunc --format="{{ .ID }}" hashicorp/vault-plugin-secrets-kv | cut -d: -f2)"
````
````suggestion
    ```shell-session
    $ SHA256="$(docker images --no-trunc --format="{{ .ID }}" PATH/TO/PLUGIN/IMAGE | cut -d: -f2)"
````

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    SHA256="$(docker images --no-trunc --format="{{ .ID }}" hashicorp/vault-plugin-secrets-kv | cut -d: -f2)"
+    ```
+
+1. Register a plugin, specifying your container image as `-oci_image`:
+    ```bash
+    vault plugin register \
+        -sha256="$SHA256" \
+        -oci_image=hashicorp/vault-plugin-secrets-kv \
+      secret kv-container
+    ```
````
````suggestion
1. Register the plugin with `vault plugin register` and specify your container image with `-oci_image`:
    ```shell-session
    $ vault plugin register \
        -sha256="$SHA256" \
        -oci_image=PATH/TO/PLUGIN/IMAGE \
      secret CONTAINER_NAME
    ```
````

Style correction: write in active voice, generalize example commands

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    ```
+
+1. Enable the plugin:
+    ```bash
+    vault secrets enable kv-container
+    ```
````
````suggestion
    ```shell-session
    vault secrets enable CONTAINER_NAME
    ```
````

fairclothjm replied:
@schavis I am curious about this pattern. It makes someone following these steps unable to simply copy and paste commands. They must also edit the `CONTAINER_NAME`. In my experience it is more preferable to not have to modify the command at all when I just want to get something running.

schavis replied:
Given that the kv plugin is just an example and not the only plugin folks may need to follow these steps for, the need to edit the line is true regardless. Using an obviously named placeholder makes the text more accessible to a broad range of readers because it makes it more obvious what information they need to provide and helps readers make connections between commands to see how the steps are related.

fairclothjm replied:
@schavis Thanks for the reply! I can definitely see your point. I wonder if a good compromise would be to use an actual shell variable in the commands and set it at the beginning? Similar to what we do with SHA256? Something like

```bash
NAME=kv-container
TAG=hashicorp/vault-plugin-secrets-kv
$ # step 1 ...
$ vault plugin register \
        -sha256="$SHA256" \
        -oci_image="$TAG" \
      secret "$NAME"
$ vault secrets enable "$NAME"
```

That way anyone that wants to directly follow along can do so without any modifactions. And others can just modify a few shell variables.

Anyway, just a thought and I will of course differ to your expertise in this area 😄

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    vault secrets enable kv-container
+    ```
+
+1. You can now interact with it like any other plugin:
+    ```shell-session
+    $ vault kv put -mount=kv-container hello subject=containers
+    Success! Data written to: kv-container/hello
+
+    $ vault kv get -mount=kv-container hello
+    ===== Data =====
+    Key        Value
+    ---        -----
+    subject    containers
+    ```
````
````suggestion

## Step 4: Test your plugin

Now that the container is registered with Vault, you should be able to interact
with it like any other plugin.

1. Write a secret to your new plugin:
   ```shell-session
   $ vault PLUGIN_NAME put -mount=my-container hello subject=containers
   Success! Data written to: my-container/hello
   ```

1. Fetch the secret you just wrote:
   ```shell-session
   $ vault PLUGIN_NAME get -mount=my-container hello
   ===== Data =====
   Key        Value
   ---        -----
   subject    containers
   ```
````

It's always good to make test steps explicit so folks know it's not part of the setup

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+Works, but does not currently support cgroup limits, and requires  passing an
+additional `--ignore-cgroups` flag to the runsc runtime.
+
+## Podman
```
If the process is as similar as we imply, we should fold the Podman steps into the guide rather than putting it at the end where it could easily be missed.

tomhjp replied:
My main hesitation here was that it felt overwhelming to present 4 different options concurrently. Would you prefer to split this into multiple smaller guides, or present it all together?

schavis replied:
The general rule of thumb is to keep things together and use tabs to separate versions of the same instructions (at least until we get variants for docs). It's better for SEO and it lets readers know their full set of options, instead of hoping they'll find a second version of the same doc with different instructions for the same thing.

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    subject    containers
+    ```
+
+### Container lifecycle and metadata
````
Everything between here and the Podman instructions should be a separate overview document since we're discussing how the plugins work rather than providing explicit instructions for how to do something.

tomhjp replied:
Oh I didn't read this before my comment above. WDYT about where the two documents belong? I think where this document currently is is a good place for reference documentation. Would you put the how-to guide alongside it or somewhere else in the tree entirely?

schavis commented on website/content/docs/plugins/containerized-plugins.mdx:
````diff
@@ -0,0 +1,253 @@
+    ---        -----
+    subject    containers
+    ```
+
````
What are some relevant next steps we can add here as a "Next steps" section? Are there other docs or tutorials we can suggest the look at for additional customization or better understanding?

tomhjp commented:
Thanks for the thorough reviews all! ❤️ I need to hit pause on this PR for a day or two, but I'll be addressing all of the feedback as soon as I can after that.

tomhjp commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+description: Running external plugins for Vault in containers.
+---
+
+# Containerized Plugins (Beta)
```
My initial aim when writing this page was to write reference documentation and that's reflected in the positioning of the page within the sidebar. However, quite a lot of it ended up being how-to. Maybe I should split it into 2 pages?

tomhjp commented on website/content/docs/plugins/containerized-plugins.mdx:
```diff
@@ -0,0 +1,253 @@
+
+### Installation
+
+1. Install a container engine, for example:
+    * [Docker](https://docs.docker.com/engine/install/)
+    * Or [rootless Docker](https://docs.docker.com/engine/security/rootless/)
```
That's a maturity question we'll answer properly when we make the feature GA.

tomhjp commented:
I've split out the tutorial-like content into a separate PR creating a brand new tutorial. When that lands, I'll link to it from these docs.

That just leaves reference documentation in this PR which I'd like to land in time for the 1.15 release if possible. I have attempted to address all comments, but may have missed some as it has been quite challenging to keep track of so many and with such large changes to the content.

tomhjp commented:
I'm happy to address more follow-up comments in another PR, but going to land this for now so the documentation exists in time for the GA release.
